File: src/main/java/no/regnskap/controller/RegnskapApiImpl.java
package no.regnskap.controller;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import no.regnskap.generated.model.Regnskap;
import no.regnskap.generated.model.Regnskapstype;
import no.regnskap.jena.ExternalUrls;
import no.regnskap.jena.JenaUtils;
import no.regnskap.mapper.RegnskapFieldsMapper;
import no.regnskap.model.Partner;
import no.regnskap.repository.ConnectionManager;
import no.regnskap.service.RegnskapService;
import no.regnskap.service.RestcallLogService;
import no.regnskap.spring.config.ProfileConditionalValues;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.stereotype.Controller;
import org.springframework.util.MimeType;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import java.util.List;


@Controller
@RestControllerAdvice
public class RegnskapApiImpl implements no.regnskap.generated.api.RegnskapApi {
    private static final Logger LOGGER = LoggerFactory.getLogger(RegnskapApiImpl.class);

    @Autowired
    private ProfileConditionalValues profileConditionalValues;

    @Autowired
    private RegnskapService regnskapService;

    @Autowired
    private RestcallLogService restcallLogService;

    @Autowired
    private ConnectionManager connectionManager;


    @Override
    public ResponseEntity<List<String>> getLog(HttpServletRequest httpServletRequest) {
        try {
            restcallLogService.logCall(httpServletRequest, "getLog");

            List<String> logList = regnskapService.getLog();

            if (logList == null || logList.size() == 0) {
                return new ResponseEntity<>(HttpStatus.NOT_FOUND);
            } else {
                return new ResponseEntity<>(logList, HttpStatus.OK);
            }
        } catch (Exception e) {
            LOGGER.error("getLog failed: ", e);
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    @Override
    @Operation(security = @SecurityRequirement(name = "basicAuth"))
    public ResponseEntity<Object> getRegnskapDeprecated(HttpServletRequest httpServletRequest, String orgNummer, Integer år, Regnskapstype regnskapstype) {
        return getRegnskap(httpServletRequest, orgNummer, år, regnskapstype);
    }

    @Override
    @Operation(security = @SecurityRequirement(name = "basicAuth"))
    public ResponseEntity<Object> getRegnskap(HttpServletRequest httpServletRequest, String orgNummer, Integer år, Regnskapstype regnskapstype) {
        try {
            restcallLogService.logCall(httpServletRequest, "getRegnskap", orgNummer);

            RegnskapFieldsMapper.RegnskapFieldIncludeMode regnskapFieldIncludeMode = RegnskapFieldsMapper.RegnskapFieldIncludeMode.DEFAULT;
            try {
                Partner partner = Partner.fromRequest(connectionManager, httpServletRequest);
                if (partner!=null && partner.isAuthorized()) {
                    regnskapFieldIncludeMode = RegnskapFieldsMapper.RegnskapFieldIncludeMode.PARTNER;
                }
            } catch (IllegalArgumentException e) {
                return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
            }

            List<Regnskap> regnskapList = regnskapService.getByOrgnr(orgNummer, null, år, regnskapstype, regnskapFieldIncludeMode);

            if (regnskapList == null || regnskapList.size() == 0) {
                return new ResponseEntity<>(HttpStatus.NOT_FOUND);
            } else {
                MimeType negotiatedMimeType = JenaUtils.negotiateMimeType(httpServletRequest.getHeader("Accept"));
                if (JenaUtils.jenaCanSerialize(negotiatedMimeType)) {
                    ExternalUrls urls = new ExternalUrls(profileConditionalValues.regnskapsregisteretUrl(),
                                                         profileConditionalValues.organizationCatalogueUrl());
                    String body = JenaUtils.modelToString(JenaUtils.createJenaResponse(regnskapList, urls), JenaUtils.mimeTypeToJenaFormat(negotiatedMimeType));
                    return ResponseEntity.ok().contentType(MediaType.asMediaType(negotiatedMimeType)).body(body);
                } else {
                    return new ResponseEntity<>(regnskapList, HttpStatus.OK);
                }
            }
        } catch (Exception e) {
            LOGGER.error("getRegnskap failed: ", e);
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    @Override
    @Operation(security = @SecurityRequirement(name = "basicAuth"))
    public ResponseEntity<Object> getRegnskapById(HttpServletRequest httpServletRequest, String orgNummer, Integer id) {
        try {
            restcallLogService.logCall(httpServletRequest, "getRegnskapById");

            RegnskapFieldsMapper.RegnskapFieldIncludeMode regnskapFieldIncludeMode = RegnskapFieldsMapper.RegnskapFieldIncludeMode.DEFAULT;
            try {
                Partner partner = Partner.fromRequest(connectionManager, httpServletRequest);
                if (partner != null && partner.isAuthorized()) {
                    regnskapFieldIncludeMode = RegnskapFieldsMapper.RegnskapFieldIncludeMode.PARTNER;
                }
            } catch (IllegalArgumentException e) {
                return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
            }

            List<Regnskap> regnskaper = regnskapService.getByOrgnr(orgNummer, id, null, null, regnskapFieldIncludeMode);
            Regnskap regnskap = (regnskaper==null || regnskaper.isEmpty()) ? null : regnskaper.get(0);

            if (regnskap == null) {
                return new ResponseEntity<>(HttpStatus.NOT_FOUND);
            } else {
                MimeType negotiatedMimeType = JenaUtils.negotiateMimeType(httpServletRequest.getHeader("Accept"));
                if (negotiatedMimeType == null) {
                    return new ResponseEntity<>(regnskap, HttpStatus.OK);
                } else {
                    ExternalUrls urls = new ExternalUrls(profileConditionalValues.regnskapsregisteretUrl(),
                                                         profileConditionalValues.organizationCatalogueUrl());
                    String body = JenaUtils.modelToString(JenaUtils.createJenaResponse(regnskap, urls), JenaUtils.mimeTypeToJenaFormat(negotiatedMimeType));
                    return ResponseEntity.ok().contentType(MediaType.asMediaType(negotiatedMimeType)).body(body);
                }
            }
        } catch (Exception e) {
            LOGGER.error("getRegnskapById failed: ", e);
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    @ExceptionHandler
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public void handleBadRequest(HttpMessageNotReadableException e) {
        LOGGER.info("Returning BAD Request: ", e);
        throw e;
    }
}

File: src/main/kotlin/no/regnskap/controller/RegnskapApiImpl.kt
package no.regnskap.controller

import no.regnskap.configuration.ProfileConditionalValues
import no.regnskap.jena.ExternalUrls
import no.regnskap.jena.JenaType
import no.regnskap.jena.acceptHeaderToJenaType
import no.regnskap.jena.createJenaResponse
import no.regnskap.service.RegnskapService
import no.regnskap.service.RestcallLogService
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping

import javax.servlet.http.HttpServletRequest

import org.springframework.web.bind.annotation.RequestMethod.GET

private val LOGGER = LoggerFactory.getLogger(RegnskapApiImpl::class.java)

@Controller
open class RegnskapApiImpl(
    private val regnskapService: RegnskapService,
    private val profileConditionalValues: ProfileConditionalValues,
    private val restcallLogService: RestcallLogService
): no.regnskap.generated.api.RegnskapApi {

    val ping: ResponseEntity<String>
        @RequestMapping(value = ["/ping"], method = [GET], produces = ["text/plain"])
        get() = ResponseEntity.ok("pong")

    val ready: ResponseEntity<*>
        @RequestMapping(value = ["/ready"], method = [GET])
        get() = ResponseEntity.ok().build<Any>()

    override fun getLog(httpServletRequest: HttpServletRequest): ResponseEntity<List<String>> {
        restcallLogService.logCall(httpServletRequest, "getLog", null);
        return ResponseEntity(regnskapService.getLog(), HttpStatus.OK)
    }

    override fun getRegnskap(httpServletRequest: HttpServletRequest, orgNummer: String, år: Int?, regnskapstype: String?): ResponseEntity<Any> =
        try {
            restcallLogService.logCall(httpServletRequest, "getRegnskap", orgNummer);
            val regnskap = regnskapService.getByOrgnr(orgNummer, år, RegnskapService.regnskapstypeToKode(regnskapstype))
            val jenaType = acceptHeaderToJenaType(httpServletRequest.getHeader("Accept"))

            val urls = ExternalUrls(
                self = profileConditionalValues.regnskapsregisteretUrl(),
                organizationCatalogue = profileConditionalValues.organizationCatalogueUrl()
            )

            if (jenaType == JenaType.NOT_JENA) {
                ResponseEntity(regnskap, HttpStatus.OK)
            } else {
                ResponseEntity(regnskap.createJenaResponse(jenaType, urls), HttpStatus.OK)
            }

        } catch (e: Exception) {
            LOGGER.error("getRegnskap failed:", e)
            ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR)
        }

    override fun getRegnskapById(httpServletRequest: HttpServletRequest, id: String): ResponseEntity<Any> =
        try {
            restcallLogService.logCall(httpServletRequest, "getRegnskapById", null);
            val regnskap = regnskapService.getById(id)
            val jenaType = acceptHeaderToJenaType(httpServletRequest.getHeader("Accept"))

            val urls = ExternalUrls(
                self = profileConditionalValues.regnskapsregisteretUrl(),
                organizationCatalogue = profileConditionalValues.organizationCatalogueUrl()
            )

            if (regnskap == null) {
                ResponseEntity(HttpStatus.NOT_FOUND)
            } else if (jenaType == JenaType.NOT_JENA) {
                ResponseEntity<Any>(regnskap, HttpStatus.OK)
            } else {
                ResponseEntity<Any>(regnskap.createJenaResponse(jenaType, urls), HttpStatus.OK)
            }

        } catch (e: Exception) {
            LOGGER.error("getRegnskapById failed:", e)
            ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR)
        }
}